### DISCUSSION TITLE
Adding text layer to a scanned PDF

### DISCUSSION THREAD

**Discussion by andrei-volkau** [pymupdf/PyMuPDF#775] [Q&A] [answered]
It would be good to know how to add a text layer to a scanned PDF.

Let's consider the following document as an example.
[test.pdf](https://github.com/pymupdf/PyMuPDF/files/5710914/test.pdf)

Raw JSON response of Amazon Textract API call:
[pdf-response.json.zip](https://github.com/pymupdf/PyMuPDF/files/5710918/pdf-response.json.zip)

**Comment by JorjMcKie** [comment, marked as answer]
Without really knowing the internals of the attached JSON info, let's assume it contains all the required text.
Then simply create a `TextWriter` object an `append` text piece by text piece
```python
tw = fitz.TextWriter(page.rect)  # need the intended page's size here
# for each text piece (a word, a string, a character, ... everything goes)
tw.append(
    pos,  # the insertion point
    text,  # the text to insert
    font=font,  # a fitz.Font(...) object
    fontsize=fontsize,
   )
# ... repeat the above with arbitrary other fonts / fontsizes, when done:
tw.writeText(page, render_mode=3,...)  # write the whole text writer as hidden (render mode 3) text.
```

**Reply by sjscotti** [reply]
Hi, I have a question.
Since some OCR systems do a segmentation of text into blocks, lines, and words (e.g., OCR-D, which also has repos on github), is there a way to include that block and line information along with the word information in the hidden text?  This information can be used to help identify reading order or separate the text from different articles for complex layouts such as newspapers.  Since block number, line number, and word number can be extracted from a pdf using `page.get_text("words")`, I would think there is a way to set those parameters when writing a pdf too.

**Comment by JorjMcKie** [comment]
@sjscotti - not quite clear what you mean:
If you do `page.get_text(...)` you will get all the same information whether OCR or not - including blocks, lines, ...
Including output from other packages is a niche request for PyMuPDF. You can develop something yourself to integrate that with the PyMuPDF output.

**Reply by sjscotti** [reply]
I implemented something much like you described above where I am using a series of `TextWriter.append` commands to place each individual word on an image-only PDF.  Since I am only supplying the 
```
    pos,  # the insertion point
    text,  # the text to insert
    font=font,  # a fitz.Font(...) object
    fontsize=fontsize,
```
info, as done in your example, when I later reopen the PDF and do a `Page.get_text("word")` command, it will return the dict with bounding box and the text of the word along with block, line and word number.  But how does it know what to use for block, line, and word number since I never input those when creating the text on the page?  But it is creating those I found out.  However, I would have grouped them differently to correspond to what the OCR program I am using detected.

**Comment by JorjMcKie** [comment]
@sjscotti 
>  But how does it know what to use for block, line, and word number

This information is generated automatically by MuPDF heuristics, when a page is read.
